### PR TITLE
auth: use workload identity when fetching secrets

### LINKED ISSUE #11: Support Workload Identity
### TL;DR

The CSI driver should use the identity of the pod when fetching secrets instead of the instance identity.

### Design

**Proposal**

Modify the plugin to fetch credentials that are compatible with [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) by default to promote least privilege permissions on secrets.

Grant the CSI driver the `serviceaccounts/token` permission in the k8s cluster and use the `pod.uid` to fetch tokens for the pod to then communicate with the `secretmanager.googleapis.com` API.

**Alternatives considered**

**Resources**

**Additional information**

### PR DESCRIPTION
**Pull request by tam7t**
Workload Identity is explained here: https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity

Supporting workload identity is needed so that IAM policies on secrets can be limited to only the workloads that use those secrets (otherwise an entire GKE cluster would share a single identity for accessing the Secret Manager API).

Fixes #11

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,6 +27,29 @@ $ ./scripts/build.sh
...
+$ gcloud iam service-accounts add-iam-policy-binding \
+    --role roles/iam.workloadIdentityUser \
+    --member "serviceAccount:$PROJECT_ID.svc.id.goog[default/mypod]" \
+    gke-workload@$PROJECT_ID.iam.gserviceaccount.com
+```
+* Create a secret that the workload identity service account can access
+```shell
+# Create a secret with 1 active version
+$ echo "foo" > secret.data
+$ gcloud secrets create test --replication-policy=automatic --data-file=secret.data
```
Comment (colinman) [inline README.md:45]: nit: `testsecret` or `mysecret` here is probably more descriptive to help make it crystal clear in all contexts

File: README.md [hunk 2]
```diff
@@ -27,6 +27,29 @@ $ ./scripts/build.sh
...
 ```
+* Setup the workload identity service account
+```shell
+# Create a service account for workload identity
+$ gcloud iam service-accounts create gke-workload
+
+# Allow "default/mypod" to act as the new service account
+$ gcloud iam service-accounts add-iam-policy-binding \
+    --role roles/iam.workloadIdentityUser \
+    --member "serviceAccount:$PROJECT_ID.svc.id.goog[default/mypod]" \
```
[the remark this answers is not in the document]
Comment (tam7t) [reply README.md:38]: sgtm!

File: README.md (hunk 1 above)
Comment (tam7t) [reply README.md:45]: sgtm!

File: README.md [hunk 3]
```diff
@@ -27,6 +27,29 @@ $ ./scripts/build.sh
 ```shell
 $ ./scripts/deploy.sh
 ```
+* Setup the workload identity service account
+```shell
+# Create a service account for workload identity
+$ gcloud iam service-accounts create gke-workload
```
[the remark this answers is not in the document]
Comment (tam7t) [reply README.md:33]: I think you would generally do 1:1 for all KSAs that need to access GCP resources. If the KSA pod is only interacting with kubernetes or other pods then I believe you would just omit the GCP SA annotation from the KSA yaml.

**Pull request merged**